## Data

For this project, I used the online wikipedia webpage (https://en.wikipedia.org/wiki/List_of_districts_and_neighborhoods_of_Los_Angeles) to get the neighborhoods of Los Angeles city.

Since I need to extract data from a webpage, I imported library of BeautifulSoup to clean the html file. I slicsed the data from the webpage contained the neighborhoods names and created a pandas dataframe with one column of Neighborhoods for further analysis. 

Then I imported nominatim to find the latitude and longitude of the corresponding neighborhood followed by inserting into the previous dataframe with column names of latitude and longitude. 


In [1]:
from bs4 import BeautifulSoup
import requests
import csv
import pandas as pd
import numpy as np
import re

# get the table's original data
source = requests.get('https://en.wikipedia.org/wiki/List_of_districts_and_neighborhoods_of_Los_Angeles').text
soup = BeautifulSoup(source,'lxml')

# get all data of the neighborhood(except ones can't find the latitude and longitude, 
# also the neighborhoods that far from the central Los Angeles) and make a list
data = []

table = soup.find_all('div', class_='div-col')

for raw_data  in  table[0].find_all('li'):
    all_data=raw_data.text
    all_data=all_data.rstrip()
    if re.search("/", all_data) or re.search(",", all_data) or re.search("Beachwood Canyon", all_data) or re.search("Holmby Hills", all_data) or re.search("NoHo Arts District", all_data) or re.search("Picfair Village", all_data) or re.search("Yucca Corridor", all_data):
        continue
    data.append(all_data)

for raw_data  in  table[1].find_all('li'):
    all_data=raw_data.text
    all_data=all_data.rstrip()
    if re.search("/", all_data) or re.search(",", all_data) or re.search("Beachwood Canyon", all_data) or re.search("Holmby Hills", all_data) or re.search("NoHo Arts District", all_data) or re.search("Picfair Village", all_data) or re.search("Yucca Corridor", all_data):
        continue
    data.append(all_data)
# creat the dataframe
data_array0=np.array(data)
data_array1 = [x.split("[")[0] for x in data_array0]

not_wanted = ["Arlington Heights","Canterbury Knolls","Del Rey", "Edendale", "Harvard Park", "Ladera", "Nichols Canyon", "Rancho Park", "Sunland","Valley Glen"]
data_array = [x.split("[")[0] for x in data_array1 if x not in not_wanted]

df = pd.DataFrame(data_array, columns ={"Neighborhood"})
df.columns = ['Neighborhood']
df['Latitude']=''
df['Longitude']=''
df.shape


(176, 3)

In [2]:
# use geopy to get the latitude and longitude of neighborhoods of Los Angeles

from geopy.geocoders import Nominatim

for x in range(len(data_array)):
    
    neighborhood_name = df['Neighborhood'][x]
    address = neighborhood_name +', CA' 
    geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(address)
    latitude = location.latitude
    longitude = location.longitude
   
    df['Latitude'][x] = latitude
    df['Longitude'][x] = longitude

df

Neighborhood Latitude Longitude
0             Angelino Heights  34.0703  -118.255
1                       Arleta  34.2413  -118.432
2                Arts District  42.3178  -83.0415
3              Atwater Village  34.1164  -118.256
4                Baldwin Hills  34.0076  -118.351
5              Baldwin Village  47.5956  -57.6407
6                Baldwin Vista  34.4302  -119.732
7              Benedict Canyon  34.0494    -118.4
8                Beverly Crest  35.4645   -80.826
9                 Beverly Glen  43.0554  -82.1759
10               Beverly Grove  53.5697  -113.402
11   Beverly Hills Post Office  53.5593  -113.031
12                Beverly Park  34.0638  -118.265
13                 Beverlywood  34.0466  -118.395
14               Boyle Heights  34.0332  -118.205
15                   Brentwood  51.0931  -114.126
16            Brentwood Circle  33.7412  -116.987
17              Brentwood Glen  29.7426  -95.7543
18         Broadway-Manchester  53.5123   -2.1696
19                   Brookside  45.3873  -63.2419
20                 Bunker Hill  50.0208  -96.3705
21               Cahuenga Pass  34.1269  -118.346
22                 Canoga Park  34.2011  -118.598
23                     Carthay  34.0561  -118.373
24              Castle Heights  45.4308  -75.6386
25             Central-Alameda  49.2667  -102.283
26                Central City  49.2769  -123.115
27                Century City  34.0556  -118.418
28                  Chatsworth  44.4561  -80.8979
29         Chesterfield Square  43.2952   -80.667
..                         ...      ...       ...
146                   Van Nuys  34.1867  -118.449
147                     Venice  54.6981  -112.147
148             Vermont Knolls  44.3734  -72.7163
149            Vermont-Slauson  33.9884   -118.29
150             Vermont Square  43.6683  -79.4128
151              Vermont Vista   34.121  -117.548
152             Victor Heights  52.3459  -106.766
153              Victoria Park   50.448  -104.612
154              Village Green   46.214  -62.9247
155         Warehouse District  35.7758  -78.6454
156              Warner Center  34.1811  -118.597
157                      Watts  51.6624  -112.091
158                 West Adams  40.8284  -89.3733
159                Westchester   45.616  -63.6745
160                   Westdale  43.2619  -79.9059
161            Western Heights  47.6178  -52.8386
162                 West Hills  45.9967  -66.6411
163                   Westlake  43.6231  -79.4835
164           West Los Angeles  34.0464  -118.448
165           Westside Village  36.7399  -78.9456
166                   Westwood  44.3214  -78.0657
167           Westwood Village  34.0595  -118.444
168            Whitley Heights  35.5556  -78.3862
169         Wholesale District  49.2375  -124.044
170                 Wilmington    33.78  -118.263
171            Wilshire Center  43.7809  -79.2326
172              Wilshire Park  35.5843  -82.6118
173             Windsor Square  34.0726  -118.321
174                   Winnetka  34.2013  -118.571
175             Woodland Hills  48.8265  -123.626

[176 rows x 3 columns]

Then I imported geopy library and foium library for making a map. There are totally 176 neighborhoods in this dataframe. I created a map of neighborhoods of Los Angeles city.

After cleaning, selecting and visualizing the dataset, I utilized Foursquare API to explore the neighborhoods of Los Angeles. Foursquare API has a massive dataset of location data. They crowdsourced their data and then we can build our dataset and add venues and explore the region that we are interested in. In my project, I requested the top 100 venus in Los Angeles region with four columns of venus name, venus catogories, venus latitude and longitude respectly. 

Before analyzing the dataset, I need to create a new dataframe including neighborhood name, neighborhood latitude, neighborhood longtitude, venus name, venus latitude, venus longtitude and venus categories. Then I inputed each time the venus was visited. Now I can group rows by neighborhood name and take the mean of the frequency of occurrence of each venus category. Then the dataframes of the top ten most common venus were created. 

After I customerize the dataset, I can apply customer segmentation on this historical data. The algorithms that I used for segmentation is K-means clustering. The number of cluster that I choosed is 5. I imported Kmeans to cluster the dataset and imported forlium to visualize the five  clusters.